### DISCUSSION TITLE
Use MaybeNullableValue in other postgres function

### DISCUSSION THREAD

**Discussion by guissalustiano** [diesel-rs/diesel#4241] [Ideas]
In postgres `to_json` function @weiznich comes with a ~magic~ [`MaybeNullableValue`](https://github.com/diesel-rs/diesel/pull/4229/files#diff-bf88b4900340a435caafd7c947e7b9a95e61ff158ef37596ab3262b3c754c45fR3702-R3714) implementation (he provides an explanation [here](https://github.com/diesel-rs/diesel/pull/4229#discussion_r1743761013))
This trait allows "overload" the `to_json` function to work as
- `to_json(e: SqlType) -> Json`
- `to_json(e: Nullable<SqlType>) -> Nullable<Json>`
Which makes their use much cleaner, without the need to call `assume_not_nullable()` and just trust in the type system.

There are multiple other functions with this same behavior listed below: if you pass null, Postgres returns null, if you don't pass null it will never return not null
- isempty
- lower_inc
- upper_inc
- lower_inf
- upper_inf
- cardinality
- array_length
- array_lower
- array_position
- array_position_with_subscript
- array_positions
- array_ndims
- array_upper

Would it be possible to generalize the `MaybeNullableValue` so we can use it in the other cases? Or would it be better to make a helper type for each one?

Thanks in advance!

**Comment by weiznich** [comment]
Got catch. I would say the correct solution is to use `MaybeNullableValue<>` for these functions as well. As far as I can see it should already be possible to just use the existing trait for that, so we shouldn't need new helper traits for that. Essentially you would transform `fn foo<T: AnyOtherTrait>(t: T) -> ReturnType` into `fn foo<T: AnyOtherTrait + MaybeNullableValue<ReturnType>(t: T) -> T::Out` and it should work.

**Reply by guissalustiano** [reply]
Awesome, I didn't understand only looking at the type signature. Thanks!

**Reply by guissalustiano** [reply]
I implemented that on https://github.com/diesel-rs/diesel/pull/4260
